### ISSUE TITLE
What kind of bug does Cannot_star report? 

### ISSUE THREAD

**Issue by bmwvvvv** [facebook/infer#1090]
infer version : v0.15.0
run cmd:  infer -a checkers --debug --no-filtering -- make

output:
script/cba_make_files/hl_0.cpp:22394: error: Cannot_star (biabduction/Tabulation.ml:655:34-41:)
  .
  22392.       BOOL   bSwitchFlag = CCmmTrans::IsSwitchOpen(wUserID);
  22393.       BOOL   bFuncFlag   = CCmmTrans::IsFuncOpen(wUserID);
  22394. >     BOOL   bStatSceneFlag = CCmmTrans::IsStatScene(wUserID);
  22395.       BOOL   bModeFlag = CCmmUtility::IsSwchOpen(wUserID);
  22396.   

script/cba_make_files/hl_0.cpp:22394: error: NULL_DEREFERENCE (biabduction/Tabulation.ml:1341:63-70:)
  [B5] pointer `CLclSingleton<CLclTrans>::s_pInstance` could be null and is dereferenced by call to `IsStatScene` at line 22394, column 32.
  22392.       BOOL   bSwitchFlag = CCmmTrans::IsSwitchOpen(wUserID);
  22393.       BOOL   bFuncFlag   = CCmmTrans::IsFuncOpen(wUserID);
  22394. >     BOOL   bStatSceneFlag = CCmmTrans::IsStatScene(wUserID);
  22395.       BOOL   bModeFlag = CCmmUtility::IsSwchOpen(wUserID);
  22396.   

Alas, why trace the infer source code line  (biabduction/Tabulation.ml:655:34-41:)
(biabduction/Tabulation.ml:1341:63-70:)

**Comment by jvillard** [comment]
"Cannot_star" indicates a bug in the analysis and is meant for infer developers. This is also why the error mentions the location in the infer source code. You're seeing developer warnings because you passed `--debug`.

**Comment by junhyung-lee38** [comment]
Dear ngorogiannis,

Thank you for your prompt reply.
It helped me a lot.

If you don't mind can I get some more advice with this brilliant tool ?

I have encountered lots of problems regarding misuse of MUTEX_LOCK (which results in race conditions).
So I really would like to detect potential errornous statements in advance before running at real situations.

Below are my questions (or naive ideas ) ...
﻿
1) Are there any plan to make 'infer' smarter by adding feature like recognizing 'pthread_mutex lock' 
    ( Because I really need this for the moment and pthread_mutex is so common in POSIX based OS like Linux ).
    I'm just wondering if there is any specific reason for not supporting this .  It would be very helpful and thankful to support this feature, I think.

2) Is it possible or are there any plans for 'infer' command-line option to provide some efficient way to train additional tokens which could be defined by user (programmer) or some standards ? 
    ( Tokens might be user-defined or C++ STL funtions like std::vector<int>::iterator type and shared variable which is defined as that type )
  ﻿Actually I got to think about above idea as Ocaml is an language for AI applications, I heard.. 

They are just naive ideas but it would be appreciated if I could have your thoughtful opinion or answers .

I look forward to hearing from you.
Best regards,
